apis/python/tests/test_indexed_dataframe.py

```diff
@@ -289,6 +294,267 @@ def make_dataframe(request):
+            "ids": [slice(None, None)],  # Indexing slot is slice-all
+            "A": [10, 11, 12, 13, 14, 15],
+            "throws": None,
+        },
+        {
+            "index_column_names": ["index1"],
+            "ids": [slice(None, 3)],  # Half-slices are not supported yet
+            "A": None,
+            "throws": ValueError,
+        },
+        {
+            "index_column_names": ["index1"],
+            "ids": [slice(1, None)],  # Half-slices are not supported yet
+            "A": None,
+            "throws": ValueError,
+        },
+        {
+            "index_column_names": ["index1"],
+            "ids": [slice(1, 5, 2)],  # Slice step must be 1 or None
+            "A": None,
+            "throws": ValueError,
+        },
+        # 1D: indexing slot is of invalid type
+        # TODO: I want to test this but Typeguard fails the test since it already knows strings are not
+        # valid until we implement
+        # https://github.com/single-cell-data/TileDB-SOMA/issues/418
+        # https://github.com/single-cell-data/TileDB-SOMA/issues/419
+        pytest.param(
+            {
+                "index_column_names": ["index1"],
+                "ids": ["nonesuch"],  # noqa
+                "A": None,
+                "throws": soma.SOMAError,
+            },
+            marks=pytest.mark.xfail,
+        ),
+        # 2D: indexing list is None
+        {
+            "index_column_names": ["index2", "index3"],
+            "ids": None,
+            "A": [10, 11, 12, 13, 14, 15],
+            "throws": None,
+        },
+        # 2D: indexing slot is None
+        {
+            "index_column_names": ["index2", "index3"],
+            "ids": [None, None],
+            "A": [10, 11, 12, 13, 14, 15],
+            "throws": None,
+        },
+        # 2D: indexing slot is int
+        {
+            "index_column_names": ["index2", "index3"],
+            "ids": [400, 0],
+            "A": [10],
+            "throws": None,
+        },
+        # 2D: indexing slot is list
+        # TODO: at present SOMAReader only accepts int dims. See also:
+        # https://github.com/single-cell-data/TileDB-SOMA/issues/418
+        # https://github.com/single-cell-data/TileDB-SOMA/issues/419
+        {
+            "index_column_names": ["index2", "index3"],
+            "ids": [[400, 600], None],
+            "A": [10, 11, 14, 15],
+            "throws": None,
+        },
+        # 3D: indexing slot is list
+        {
+            "index_column_names": ["index2", "index3", "index4"],
+            "ids": [[400, 600], None, None],
+            "A": [10, 11, 14, 15],
+            "throws": None,
+        },
+        # 3D: indexing slot is mixed
+        {
+            "index_column_names": ["index2", "index3", "index4"],
+            "ids": [range(400, 600), None, np.asarray([2000, 9999])],
+            "A": [11],
+            "throws": None,
+        },
+    ],
+)
+def test_indexing(tmp_path, io):
```

Comment:
name suggestion:  `test_read_indexing`.   We will need a write indexing test as well.